include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h

```diff
@@ -0,0 +1,547 @@
+{
+    using _BinHashType = typename oneapi::dpl::__internal::__custom_range_binhash<_Range>;
+    _BinHashType __bin_hash;
+
+    using __boundary_type = typename _BinHashType::__boundary_type;
+
+    __binhash_SLM_wrapper(_BinHashType __bin_hash) : __bin_hash(__bin_hash) {}
+
+    template <typename _T2>
+    inline ::std::uint32_t
+    get_bin(_T2&& __value) const
+    {
+        return __bin_hash.get_bin(::std::forward<_T2>(__value));
+    }
+
+    template <typename _T2>
+    inline bool
+    is_valid(_T2&& __value) const
+    {
+        return __bin_hash.is_valid(::std::forward<_T2>(__value));
+    }
+
+    inline ::std::size_t
+    get_required_SLM_memory()
+    {
+        return sizeof(__boundary_type) * __bin_hash.__boundaries.size();
+    }
+
+    inline void
+    init_SLM_memory(void* __boost_mem, const sycl::nd_item<1>& __self_item) const
+    {
+        ::std::uint32_t __gSize = __self_item.get_local_range()[0];
+        ::std::uint32_t __self_lidx = __self_item.get_local_id(0);
+        ::std::uint8_t __factor = oneapi::dpl::__internal::__dpl_ceiling_div(__bin_hash.__boundaries.size(), __gSize);
+        ::std::uint8_t __k;
+        __boundary_type* __d_boundaries = static_cast<__boundary_type*>(__boost_mem);
+        for (__k = 0; __k < __factor - 1; __k++)
+        {
+            __d_boundaries[__gSize * __k + __self_lidx] = __bin_hash.__boundaries[__gSize * __k + __self_lidx];
+        }
+        // residual
+        if (__gSize * __k + __self_lidx < __bin_hash.__boundaries.size())
+        {
+            __d_boundaries[__gSize * __k + __self_lidx] = __bin_hash.__boundaries[__gSize * __k + __self_lidx];
+        }
+    }
+
+    template <typename _T2>
+    ::std::uint32_t inline get_bin(_T2&& __value, void* __boost_mem) const
+    {
+        __boundary_type* __d_boundaries = static_cast<__boundary_type*>(__boost_mem);
+        return (::std::upper_bound(__d_boundaries, __d_boundaries + __bin_hash.__boundaries.size(),
+                                   ::std::forward<_T2>(__value)) -
+                __d_boundaries) -
+               1;
+    }
+
+    template <typename _T2>
+    bool inline is_valid(const _T2& __value, void* __boost_mem) const
+    {
+        __boundary_type* __d_boundaries = static_cast<__boundary_type*>(__boost_mem);
+        return (__value >= __d_boundaries[0]) && (__value < __d_boundaries[__bin_hash.__boundaries.size()]);
+    }
+};
+
+template <typename _HistAccessor, typename _OffsetT, typename _Size>
+inline void
+__clear_wglocal_histograms(const _HistAccessor& __local_histogram, const _OffsetT& __offset, const _Size& __num_bins,
+                           const sycl::nd_item<1>& __self_item)
+{
+    ::std::uint32_t __gSize = __self_item.get_local_range()[0];
+    ::std::uint32_t __self_lidx = __self_item.get_local_id(0);
+    ::std::uint8_t __factor = oneapi::dpl::__internal::__dpl_ceiling_div(__num_bins, __gSize);
+    ::std::uint8_t __k;
```

Comment:
uninitialized variable

Reply:
It is initialized in the for loop directly below, but I can move the initialization up a line.  Its a safety vs clarity choice I think.

Reply:
Usually,  a code analyzing tool treats a variable definition without initialization as an issue....

Reply:
Thanks, I've edited this to initialize on definition.